scripts/prep-upgrade

```diff
@@ -0,0 +1,73 @@
+#!/bin/bash
+
+set -uo pipefail
+
+# This script ensures the pre-requisites are in place to perform a template
+# swap for securedrop-workstation templates from Debian Stretch to Buster.
+# To allow the template of an AppVM to be changed, the following two
+# conditions must be met:
+# 1. The AppVM must be powered off
+# 2. The AppVM must not be a DispVM template that used as the default DispVM
+#    for an AppVM, nor the system default DispVM.
+#
+# The VMs with managed templates are as follows:
+# sd-svs, sd-svs-disp, sd-export-usb-dvm, sd-gpg
+#
+# We check that the VM exists and that the template on which it is based
+# is Stretch-based. Then we shutdown AppVMs, and delete AppVMs that are
+# DispVM templates.
+
+# sd-svs, we simply shutdown the machine as we want to preserve the data
+qvm-check sd-svs > /dev/null 2>&1
+if [[ $? -eq 0 ]]; then
+	BASE_TEMPLATE=$(qvm-prefs sd-svs template)
+
+	if [[ ! $BASE_TEMPLATE =~ "buster" ]]; then
+		qvm-check --running sd-svs > /dev/null 2>&1
+		if [[ $? -eq 0 ]]; then
+			qvm-shutdown sd-svs && sleep 5
+		fi
+	fi
+fi
+
+# For sd-svs-disp and sd-export-usb-dvm, DispVM templates. We can delete both
+# VMs since they contain no persistent data. The installer will re-create them
+# as part of the provisioning process.
+# We set the default DispVM to empty string to ensure nothing is opened in an
+# insecure (unmanaged or not yet updated) or networked vm, until the
+# provisioning process runs again and sets that value to sd-svs-disp
+qvm-check sd-svs-disp > /dev/null 2>&1
+if [[ $? -eq 0 ]]; then
+	BASE_TEMPLATE=$(qvm-prefs sd-svs-disp template)
+
+	if [[ ! $BASE_TEMPLATE =~ "buster" ]]; then
+    qubes-prefs default_dispvm ''
+		qvm-shutdown sd-svs-disp
+		sleep 5
+    qvm-remove -f sd-svs-disp
+	fi
+fi
+
+qvm-check sd-export-usb-dvm > /dev/null 2>&1
+if [[ $? -eq 0 ]]; then
+	BASE_TEMPLATE=$(qvm-prefs sd-export-usb-dvm template)
+
+	if [[ ! $BASE_TEMPLATE =~ "buster" ]]; then
+		qvm-shutdown sd-export-usb-dvm
+		qvm-shutdown sd-export-usb
+		sleep 5
+    qvm-remove -f sd-export-usb
+    qvm-remove -f sd-export-usb-dvm
+	fi
+fi
+
+# Finally for sd-gpg, we simply shutdown the machine
+qvm-check sd-gpg > /dev/null 2>&1
+if [[ $? -eq 0 ]]; then
+	BASE_TEMPLATE=$(qvm-prefs sd-gpg template)
+
+	if [[ ! $BASE_TEMPLATE =~ "buster" ]]; then
+		qvm-shutdown sd-gpg
+		sleep 5
+	fi
+fi
```

Comment:
At the end of this script seems to be the appropriate time to clean up the old VMs. That, or in a subsequent Salt task. Hardcoding the VMs to remove seems to be what we want. This Python one-liner describes the logic:

```
python3 -c 'from qubesadmin import Qubes; q = Qubes(); sdw_vms = [x.name for x in q.domains \
            if "sd-workstation" in x.tags and x.klass == "TemplateVM" \
            and "buster" not in x.name \
            and not x.installed_by_rpm ]; print("\n".join(sdw_vms))'
```

I _don't_ recommend we use that convoluted one-liner, although it does illustrate how useful the Admin API can be for use cases like this. For now, we can simply hardcode the list of "old" VMs in a Salt list and ensure they're gone.

Reply:
Unfortunately, we cannot set them as part of this script, because the old (stretch) templates are still being used: we need to wait for the state of the templates to be applies, so that the stretch templates are no longer being used by an appvm